Feature/MockRepositoryTest

HYK97 opened the pull request:
## 개요
Mock_Repostiory를 실제 DB_Repostitory와 유사하도록 Delay 기능 및 Exception Throw 기능을 추가

## 작업사항
- aop를 통해서 *Repository 가 붙은 클래스들 전체에 (Random) Delay Advice가 적용되도록 했다.
- 프록시 호출 시 실제 타깃이 실행되기 전에 RandomMethod를 통해서 Exception을 발생시킬지 말지 결정한다. Exception이 발생한다면 최대 5회까지 retry 후 throw 하도록 설계

HYK97 committed:
Feat : MockRepository에 delay 랜덤 예외 던지기 기능 추가 1차

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/aop/AspectConfig.java:
```diff
@@ -1,11 +1,107 @@
+
+
+    @Around("repository()")
+    public Object throwAdvice(ProceedingJoinPoint point) throws Throwable {
```
두 Aspect를 분리하는게 좋아보여요
- A: 예외 발생 시 재시도하는 로직
- B: 랜덤하게 예외 발생시키는 로직

A는 실제 repo를 구현했을 때 여전히 사용할 수 있는 유효한 Aspect이고
B는 mock 용으로만 사용할 Aspect이기 때문에 껐다 별개로 켰다 할 수 있는게 좋아요.

두 Aspect간의 순서도 생각해보고, 순서를 어떻게 조절할 수 있는지도 생각해보면 좋겠네요.

HYK97 replied:
1 랜덤하게 예외 발생시키는 로직 -> 2 예외 발생 시 재시도하는 로직
1번 advice에서 랜덤하게 throw 한 후 -> 2번 advice에서 catch 하는 방법을 모르겠습니다.
제 생각에는 proceed(target) 실행 후 proceed 중 발생한 throw를 처리하기 때문에 이전 advice에서 던진 예외는 별개로 다음 advice는 이전 advice의 throw를 catch 하는 게 아니라 마찬가지로 proceed 실행 후 내부에서 발행한 throw를 처리하기 때문에 이전 advice의 throw는 상관없기 때문에 재시도 로직을 실행하지 않고 이전 advice의 예외를 던지며 프로그램이 종료됩니다.
그래서 제 생각에는 예외를 발생시키는 로직은 aop를 적용시키기 힘들기 때문에 repo 내부에 있어야 할 것 같은데 어떻게 해결해야 할까요?

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/aop/AspectConfig.java:
```diff
@@ -1,11 +1,107 @@
+            } catch (TimeoutException | IOException e) {
+                log.info("예외 발생 재시도 카운트 ={} , 예외명 ={}", count, e.getMessage());
+                if (count >= 4) {
+                    throw new RuntimeException(e);
```
RuntimeException도 좋지만, RetryLimitExceededException extends RuntimeException 같은 타입을 만들어서 던지는 것도 좋겠네요

HYK97 replied:
수정했습니다.!

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/aop/AspectConfig.java:
```diff
@@ -1,11 +1,107 @@
+    private  int minPercent;
+    private  Predicate<Integer> frequency;            ;
+
+    public AspectConfig(int maxPercent, int minPercent, Predicate<Integer> frequency) {
```
코드를 처음 보는 사람 입장에서 이 클래스만 놓고 봤을 때, 무엇을 위한 percent인지, frequency인지가 드러나지 않아요.
이 클래스 명을 조금 더 구체적으로 짓고, 이와 무관한 로직은 별개의 Aspect 클래스로 분리하는게 좋아보입니다

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/aop/AspectConfig.java:
```diff
@@ -1,11 +1,107 @@
+    }
+
+    private int getRandomNumber(int min, int max) {
+        return ThreadLocalRandom.current().nextInt(min, max + 1);
```
RandomNumber 는 보통 min/max 범위로 조건을 주진 않아요. 어차피 랜덤하게 생성되기 때문이에, threshold 값 하나만 보고 조건을 걸어도 결과는 동일합니다.

HYK97 replied:
`threshold 값 하나만 보고 조건을 걸어도 결과는 동일합니다.` 이부분이 잘이해가안가네요 ㅜㅜ !
이 메서드는 세군데에 사용되는데

1. 20~100사이의 딜레이를 줄 때 min 20 max 100
2. 0,1,2로 랜덤 예외 발생시킬 때 min 0 max 2
3. 1퍼센트의 확률 0~100로 예외를 발생시킬 때 min 0 max 99

중복제거를 위해서 이처럼 정의했는데 다른 방법으로 만들어야 할까요??

f-lab-joon replied:
아 이해했습니다. ㅎㅎ 그렇다면 괜찮습니다. 
Aspect에 국한된 메소드는 아니므로 유틸 클래스 static method로 빼는게 낫겠네요. 
min / max 역전 케이스에 대해 validation 하나 추가하면 좋을 것 같습니다. 
이럴 때 자세한 에러 메시지로 IllegalArgumentException 던져주면 나중에 디버깅 할 때 은근 도움 되는 경우가 있어요

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/config/Config.java:
```diff
@@ -0,0 +1,22 @@
+                return false;
+            }
+        };
+        return new AspectConfig(100,0,predicate);
```
파라미터 percentage 하나만 남기고 이 값을 1로 주고 확률을 계산해도 동일한 로직을 구현할 수 있을 것으로 보여요. 
단순히 몇%의 확률로 적용을 하는데 있어서 불필요하게 복잡한 파라미터를 받는건 아닌지 생각해보아요

f-lab-joon replied:
이 코멘트 스킵해주세요

f-lab-joon commented on src/test/java/com/jiyong/commerce/item/config/TestConfig.java:
```diff
@@ -0,0 +1,19 @@
+    @Bean
+    public AspectConfig testAspect() {
+        Predicate<Integer> predicate = (i)->true;
+        return new AspectConfig(100,0,predicate);
```
람다는 그냥 inline으로 써도 됩니다. 
재사용되지 않는다면 변수명, 타입을 안써줘도 돼요

HYK97 replied:
수정했습니다.!

HYK97 committed:
Refactor : 2차 수정 Aspect 분리 및 랜덤 함수 간소화

기존의 AspectConfig의 책임을 두가지로 분리함  DBConnectRetryAspect(재시도 기능),MockDelayAndThrowExceptionAspect(mock 객체 딜레이 및 예외던지기)

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/util/AspectUtils.java:
```diff
@@ -0,0 +1,21 @@
+
+public class AspectUtils {
+
+    public static void sleep(int delay) {
```
sleep과 Random number는 Aspect에 국한된 유틸이 아니니, AspectUtils보다는 다른 제네럴한 네임이 낫겠네요

HYK97 replied:
수정했습니다!

HYK97 committed:
Refactor : AspectUtils -> CommonUtils 네이밍 변경 및 getRandomNumber 메서드 min max 역전 예외 추가

jiwoojiwoo commented on src/main/java/com/jiyong/commerce/common/aop/MockDelayAndThrowExceptionAspect.java:
```diff
@@ -0,0 +1,58 @@
+
+    private final Predicate<Integer> frequency;
+
+    @Around("com.jiyong.commerce.common.aop.Pointcuts.repository()")
```
해당 aspect의 경우 pointcut을 memoryRepository() 변경해주셔야 할듯합니다.
그리고 pointcut 선언도  MemoryRepository() -> memoryRepository() 로 변경 부탁드려요.

HYK97 replied:
수정했습니다..!

jiwoojiwoo commented on src/main/java/com/jiyong/commerce/common/annotation/Retryable.java:
```diff
@@ -0,0 +1,13 @@
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Retryable {
```
spring에서 제공하는 어노테이션중 @Retryable 어노테이션이 존재하여 이름을 변경하는 것이 낫지 않을까요? 
(org.springframework.retry.annotation.Retryable)

HYK97 replied:
오.. 이미 있는 어노테이션 인지는 몰랐네요!!.. Retryable이 직관적인 이름이라고 생각해서 사용했는데 혹시 겹치면 안 좋은 부분이 있을까요 ?? 단점이 있다면 앞에 Custom을 붙여서 사용해야겠네요!

jiwoojiwoo replied:
앗 저는 중복되면 혼동이 될 수 있다고 생각하여 말씀드렸는데, 별다른 단점이 없다면 상관없겠네요.
@f-lab-joon  멘토님, annotation 선언할 때 중복되어도 상관없을까요?

HYK97 committed:
Feat : Logging Aspect 추가 및 잘못 지정된 pointcut 수정

#Loggin Aspect 추가
 - controller service repository pointcut 추가
 - LogTrace 추가
 - Config 파일 Bean 설정 변경

#MockDelayAndThrowExceptionAspect @Around pointcut 수정

HYK97 committed:
Refactor : Service And Repository update delete search 추가 및 네이밍 변경

#Service item 수정 , 삭제 , 단건 조회 추가 및 itemList 및 findByItemName 네이밍 변경

#Repository item 수정 , 삭제 , id로 검색 추가 및 insertItem itemList 네이밍 변경

HYK97 committed:
Feat : ItemDto 추가 , item Model 구조 변경 및 ItemController 추가

#Dto는 Response 와 Request 를 innerClass로 캡슐화

#ItemController crud 추가

f-lab-joon commented on src/main/java/com/jiyong/commerce/common/aop/DBConnectRetryAspect.java:
```diff
@@ -0,0 +1,61 @@
+                return proceed;
+            } catch (RuntimeException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof TimeoutException || cause instanceof IOException) {
```
어떤 Exception Type일 때 retry 할지도 Retryable annotation의 속성으로 받아 설정 가능하게 빼는 방법도 있습니다. FYI

f-lab-joon reviewed (approved):
굳
